ga4gh/gff3Parser.py

```diff
@@ -0,0 +1,321 @@
+    if _encodeAttrRe.search(v):
+        return urllib.quote(v)
+    else:
+        return v
+
+
+class Feature(object):
+    """
+    One feature notation from a GFF3, missing attribute, as code by `.' in
+    the file are stored as None.
+    """
+
+    __slots__ = ("seqname", "source", "type", "start", "end", "score",
+                 "strand", "frame", "attributes", "gff3Set", "lineNumber",
+                 "parents", "children")
+
+    def __init__(self, seqname, source, type, start, end, score, strand,
+                 frame, attributes, gff3Set, lineNumber=None):
+        """
+        The attributes field is a dict of lists/tupes as attributes maybe
+        multi-valued.
+        """
+        self.seqname = seqname
+        self.source = source
+        self.type = type
+        self.start = start
+        self.end = end
+        self.score = score
+        self.strand = strand
+        self.frame = frame
+        self.attributes = copy.deepcopy(attributes)
+        self.gff3Set = gff3Set
+        self.lineNumber = lineNumber
+        self.parents = set()
+        self.children = set()
+
+    @staticmethod
+    def __dotIfNone(val):
+        return val if val is not None else '.'
+
+    def __attributeStr(self, name):
+        """
+        Return name=value for a single attribute
+        """
+        return "{}={}".format(
+            _encodeAttr(name),
+            ",".join([_encodeAttr(v) for v in self.attributes[name]]))
+
+    def __attributeStrs(self):
+        """
+        Return name=value, semi-colon-separated string for attributes,
+        including url-style quoting
+        """
+        return ";".join([self.__attributeStr(name)
+                         for name in self.attributes.iterkeys()])
+
+    def __str__(self):
+        """
+        Return the object as a valid GFF3 record line.
+        """
+        return "\t".join([self.seqname, self.source, self.type,
+                          str(self.start), str(self.end),
+                          self.__dotIfNone(self.score),
+                          self.__dotIfNone(self.strand),
+                          self.__dotIfNone(self.frame),
+                          self.__attributeStrs()])
+
+    @property
+    def featureId(self):
+        """
+        ID attribute or None if records doesn't have an ID
+        """
+        featId = self.attributes.get("ID")
+        if featId is not None:
+            featId = featId[0]
+        return featId
+
+    def getRequiredAttribute(self, name):
+        "get a require attributes list of values, error if not found"
+        values = self.attributes.get(name)
+        if values is None:
+            raise GFF3Exception(
+                "required attribute not found: {}".format(name),
+                self.gff3Set.fileName, self.lineNumber)
+        return values
+
+
+class Gff3Set(object):
+    """
+    A set of GFF3 sequence annotations
+    """
+    def __init__(self, fileName=None):
+        self.fileName = fileName
```

Comment:
I don't think it would take particularly long to make this change. Uniformity of style is essential in a large project. All other (non-trivial) classes use protected attributes.